### PR TITLE
Brings back original support for YAML now using client calls to the REST API

### PR DESCRIPTION
**Pull request by samredai**
### Summary

This brings back an original feature of DataJunction created by @betodealmeida where users can define the entire graph using a collection of local YAML files. This also includes a top level `dj.yaml` that signals the root of the project directory and allows adding a small amount of project metadata like a name and some defaults.

Currently this only covers deploying projects but I'm hoping to follow up with an export functionality that lets you export a given namespace to a local directory of YAML files.

With a local DJ server running, you can try the following while in a DJ project directory, i.e. the examples directory added as part of this PR.
```py
from datajunction import DJBuilder
from datajunction.compile import Project

dj = DJBuilder(uri="http://localhost:8000")

project = Project.load_current()
compiled_project = project.compile()
compiled_project.validate(client=dj)
compiled_project.deploy(client=dj)
```

https://github.com/DataJunction/dj/assets/43911210/76d78f62-7429-476e-8724-f021bca1641b

### Test Plan

Wrote tests and added an example project in the python client tests directory

- [ ] PR has an associated issue: #742 
- [ ] `make check` passes
- [ ] `make test` shows 100% unit test coverage

### CODE REVIEW

File: datajunction-clients/python/tests/examples/dj.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: My DJ Project
+prefix: users.me
+mode: published
+priority:
```
[the remark this answers is not in the document]
Comment (samredai) [reply datajunction-clients/python/tests/examples/dj.yaml:4]: It's less about a priority during compiling the project than it is about deploying them, how about deploy-priority?

File: datajunction-clients/python/datajunction/compile.py [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+
+class Project(BaseModel):
+    """
+    A project configuration
+    """
+
+    name: str
+    prefix: str
+    tags: Optional[List[Tag]]
+    # nodes: List[str]
```
[the remark this answers is not in the document]
Comment (samredai) [reply datajunction-clients/python/datajunction/compile.py]: Removed!

**Comment by shangyian** [comment]
Niiice, thanks for bringing this back! 🎉 

I've been playing with this and noticed one issue - it doesn't work from a Jupyter notebook since Jupyter runs its own event loop and this setup uses asyncio to do the same. This is the error: `RuntimeError: asyncio.run() cannot be called from a running event loop`. 

Do you think it's worth having asyncio call `create_task` so that it's using the current event loop and will work for Jupyter notebooks? Something like this:
```
task = asyncio.create_task(
    load_node_configs(
        repository=Path(project.root_path),
        priority=project.priority,
    ),
    name="load_node_configs",
)
await asyncio.wait(task)
definitions = task.result()
```

**Review by shangyian — COMMENTED** [review]
Niiice, thanks for bringing this back! 🎉 

I've been playing with this and noticed one issue - it doesn't work from a Jupyter notebook since Jupyter runs its own event loop and this setup uses asyncio to do the same. This is the error: `RuntimeError: asyncio.run() cannot be called from a running event loop`. 

Do you think it's worth having asyncio call `create_task` so that it's using the current event loop and will work for Jupyter notebooks? Something like this:
```python
task = asyncio.create_task(
    load_node_configs(
        repository=Path(project.root_path),
        priority=project.priority,
    ),
    name="load_node_configs",
)
await asyncio.wait(task)
definitions = task.result()
```

**Comment by samredai** [comment]
@shangyian great catch! I completely forgot about this async issue with notebooks. Let me try your solution and update the PR.

**Comment by samredai** [comment]
@shangyian I updated the PR to make it work for notebooks. There's some formatting changes we can make to the rich table to make the output from larger deployments more readable in a notebook cell but that might be better as a follow-up PR.
https://github.com/DataJunction/dj/assets/43911210/104dbf66-4791-4114-bbed-09dd36f2742e

File: datajunction-clients/python/tests/examples/dj.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (samredai) [reply datajunction-clients/python/tests/examples/dj.yaml]: Went with a `build:` section that can be expanded more as we add more levers.
```yaml
build:
  priority:
  - foo.bar
```

**Review by shangyian — APPROVED** [review]

**Pull request merged**